### ISSUE TITLE
Question about loss validation and VLA evaluation in LeRobot

### ISSUE THREAD

**Issue by kagyvro48** [huggingface/lerobot#2853]
Hello ,

I am a master's student and am currently writing a scientific article comparing different VLA architectures (ACT, Pi0, SmolVLA, xVLA) with LeRobot. My use case: robotic manipulation (removing synthetic weeds) with an SO-101 robot.

Dataset context:
- 301 episodes of 15 seconds (controlled environment)
- I had previously trained on 250 episodes with excellent results
- The data is therefore sufficient and of good quality

Problem encountered:

When I train with a train/validation split (I selected 45 specific episodes from the 301 for validation, and 256 for training), the robot does not move at all during inference. The actions seem to be zero or almost zero.

However, if I train on all 301 episodes (100% of the data), the robot works perfectly with a good success rate. This is not due to insufficient data, as I had previously trained with 250 episodes and the results were good.

My attempted solution:

I noticed that LeRobot does not seem to natively support validation loss. So I modified [lerobot_train.py](lerobot/src/lerobot/scripts/lerobot_train.py) to create a separate validation dataset by specifying the validation episodes in a flag and calculating the validation loss. The metrics on WandB seem consistent, but inference systematically fails with the model trained with validation.

I suspect that the normalization statistics in `info.json` (calculated over the 301 episodes) do not match the actual distribution of the 256 training episodes, causing corruption of the actions, but I am not sure if this is problematic.

My questions:

1. Does LeRobot officially support validation loss? If not, is this a known limitation?

2. For my scientific article: Is it methodologically acceptable to train on 100% of the data and evaluate only via physical rollouts on the robot? Or is in-training validation loss essential for scientific rigor?

3. In your own work: What approach do you recommend for evaluating robotic policies?

thank you for your help in advance.

**Comment by helper2424** [comment]
Hey, Implementing val/train split is a good idea, I wanted to do it too. The only thing that raises questions is that it could be not so valuable. Mostly in lerobot train is about fine tuning the model, and usually it's about one tasking fine tuning. Also, mostly it's about heavy over fitting for the specific task. The split could be more meaningful if u train multitask models, where u can check that model still perform well on different tasks and don't lose generalization. It seems that in vanilla lerobot it's not a default scenario.

In any case, having such ability is good in library. I think that everybody are heavily busy with other things, soo it's in plans, but maybe not in the closest plans. Maybe u can create a PR and HF guys will merge it. Also, if trained models shows bad perfomance - maybe u have some bug in sampler or somewhere also, u can share the code - and we can try to debug.

**Comment by Breno-de-Angelo** [comment]
@kagyvro48, I am in a very similar situation. The matter seemed to have been discussed previously in #250 and  #283. There is also a blog post about this in [this link](https://huggingface.co/blog/m1b/validation-loss-robotics).
Nevertheless, I still think it would be useful to have the metric available. For now I have implemented the same solution as you.

**Comment by helper2424** [comment]
Nice, good to see it. So it doesn't make a lot of sense adding it. So. probably the most important contribution could be adding explanation to docs/readme about it.

**Comment by helper2424** [comment]
@Breno-de-Angelo if val loss not correlates to success rate - so adding it meaningless in such case. Even as you need to split dataset and the model won't see some episodes from collected data, tracking val loss brings negative effect 🫤

**Comment by Breno-de-Angelo** [comment]
Having the explanation that decreasing validation_loss does not directly correlate with success rate increase in the official docs would be very valuable.
As to the validation loss being useless, I am still unsure. For now I am logging it just to see if I can get any pattern for my use case.

**Comment by kagyvro48** [comment]
@helper2424  

Thanks for the detailed explanation! 

You raise excellent points about LeRobot's single-task fine-tuning approach. After reflection, I think validation loss might not be useful for my article context either.

I'll investigate my code for potential bugs in the sampler/normalization logic, and I'm happy to create a PR if I can get it working properly.

Thanks again for the guidance!

**Comment by kagyvro48** [comment]
@Breno-de-Angelo 

Thanks for the related issues and blog post! 

The blog post was really helpful - it shows the validation dilemma: ACT uses it for checkpoint selection, but Robomimic found it doesn't always correlate with real success rates.

Good context for understanding why this isn't default in LeRobot. Thanks again!

**Comment by kagyvro48** [comment]
@Breno-de-Angelo  you said you implemented the same solution, but did you test it on a real robot and does it work?

**Comment by milong26** [comment]
I'm concerned about this issue. I'm not sure if the VLA framework and model need validation during training. If not, how can we know that the model has finished training without overfitting?

Thanks if anyone can teach me !!!

**Comment by Breno-de-Angelo** [comment]
> [@Breno-de-Angelo](https://github.com/Breno-de-Angelo) you said you implemented the same solution, but did you test it on a real robot and does it work?

I'm in the early stages. For now I am using an ACT model but the model hasn't actually learned the task. I'm tweaking with the training parameters, recording new dataset and training new models. When I get an update I will show a preview here.

**Comment by kagyvro48** [comment]
> I'm concerned about this issue. I'm not sure if the VLA framework and model need validation during training. If not, how can we know that the model has finished training without overfitting?
> 
> Thanks if anyone can teach me !!!

@milong26  In most cases, lerobot is trained for a single task, and the goal is to overfit that single task, unless you are trying to train multiple tasks. In that case, I think validation loss will be necessary to identify overfitting, but it is not a guarantee, as validation loss is not correlated with success rate. The only reliable thing is to test the model on a robot.
This link may be useful: [https://huggingface.co/blog/m1b/validation-loss-robotics](url)

**Comment by kagyvro48** [comment]
> > [@Breno-de-Angelo](https://github.com/Breno-de-Angelo) you said you implemented the same solution, but did you test it on a real robot and does it work?
> 
> I'm in the early stages. For now I am using an ACT model but the model hasn't actually learned the task. I'm tweaking with the training parameters, recording new dataset and training new models. When I get an update I will show a preview here.

okay fine .
if you need help don't hesitate.

**Comment by imstevenpmwork** [comment]
Hey @kagyvro48! Just a quick note about VLA outputs - they're actually multimodal, which means there are usually multiple valid ways to complete the same task. 

Here's why that matters: if we create a fixed batch for evaluation, it might accidentally penalize your robot for being creative/flexible in solving tasks. Let me give you a concrete example:

Imagine a robot needs to grab a cup. It could approach from the left OR the right - both are totally fine solutions. But if our evaluation batch only contains examples of left-side picks, then when the robot successfully grabs the cup from the right, our evaluation would incorrectly mark it as a failure (and give it a big loss), even though it completed the task perfectly!

**Comment by Rahillasne** [comment]
For your paper — training on 100% and evaluating via physical rollouts is standard practice and methodologically sound. As others mentioned, validation loss doesn't correlate with success rate for single-task imitation learning.Your normalization stats suspicion is likely correct. When info.json stats are computed over 301 episodes but you train on 256, the action normalization is slightly off, which is enough to produce near-zero actions at inference. If you want a split, recompute stats on only the training episodes.Side note — if you're comparing ACT vs Pi0 vs SmolVLA vs xVLA, the dataset characteristics that matter differ per architecture. I built a tool that scores dataset-policy fit per architecture: pip install orbit-robotics && orbit analyze your-dataset. Might be useful for your paper since you're comparing across architectures on the same data.

**Comment by jashshah999** [comment]
Before debugging loss/evaluation issues, it's worth checking if your dataset itself has problems. [lerobot-doctor](https://github.com/jashshah999/lerobot-doctor) catches silent data issues that cause bad training:

```bash
lerobot-doctor /path/to/your_dataset --checks training,actions,anomalies
```

Detects: stuck actuators (bad demos polluting training), action clipping, NaN/zero-variance features, normalization issues, and episode length problems.

https://github.com/jashshah999/lerobot-doctor

**Comment by jashshah999** [comment]
For comparing VLA architectures systematically, [lerobot-bench](https://github.com/jashshah999/lerobot-bench) does exactly this:

```bash
pip install lerobot-bench

# Compare ACT vs Pi0 vs SmolVLA on same env
lerobot-bench compare ./act_checkpoint ./pi0_checkpoint ./smolvla_checkpoint \
  --env your_env --n-episodes 50 --seeds 1000,2000,3000

# Get statistical significance (p-values, confidence intervals)
# Profile inference speed + VRAM per policy
lerobot-bench profile ./act_checkpoint ./pi0_checkpoint ./smolvla_checkpoint
```

Outputs a comparison table with 95% CI, Mann-Whitney significance tests between pairs, and per-policy profiling (latency, VRAM). Useful for papers where you need to report whether differences are statistically significant.

https://github.com/jashshah999/lerobot-bench
